libraries/persistent_store/src/buffer.rs

```diff
@@ -0,0 +1,668 @@
+/// operations for inspection or to corrupt the operation.
+#[derive(Clone)]
+pub struct BufferStorage {
+    /// Content of the storage.
+    storage: Box<[u8]>,
+
+    /// Options of the storage.
+    options: BufferOptions,
+
+    /// Number of times a word was written since the last time its page was erased.
+    word_writes: Box<[usize]>,
+
+    /// Number of times a page was erased.
+    page_erases: Box<[usize]>,
+
+    /// Interruption state.
+    interruption: Interruption,
+}
+
+/// Options of a buffer storage.
+#[derive(Clone, Debug)]
+pub struct BufferOptions {
+    /// Size of a word in bytes.
+    pub word_size: usize,
+
+    /// Size of a page in bytes.
+    pub page_size: usize,
+
+    /// How many times a word can be written between page erase cycles.
+    pub max_word_writes: usize,
+
+    /// How many times a page can be erased.
+    pub max_page_erases: usize,
+
+    /// Whether bits cannot be written from 0 to 1.
+    pub strict_write: bool,
+}
+
+/// Corrupts a slice given actual and expected value.
+///
+/// A corruption function is called exactly once and takes 2 arguments:
+/// - A mutable slice representing the storage before the interrupted operation.
+/// - A shared slice representing what the storage would have been if the operation was not
+///   interrupted.
+///
+/// The corruption function may flip an arbitrary number of bits in the mutable slice, but may only
+/// flip bits that differ between both slices.
+pub type BufferCorruptFunction<'a> = Box<dyn FnOnce(&mut [u8], &[u8]) + 'a>;
+
+impl BufferStorage {
+    /// Creates a buffer storage.
+    ///
+    /// # Panics
+    ///
+    /// The following preconditions must hold:
+    /// - `options.word_size` must be a power of two.
+    /// - `options.page_size` must be a power of two.
+    /// - `options.page_size` must be word-aligned.
+    /// - `storage.len()` must be page-aligned.
+    pub fn new(storage: Box<[u8]>, options: BufferOptions) -> BufferStorage {
+        assert!(options.word_size.is_power_of_two());
+        assert!(options.page_size.is_power_of_two());
+        let num_words = storage.len() / options.word_size;
+        let num_pages = storage.len() / options.page_size;
+        let buffer = BufferStorage {
+            storage,
+            options,
+            word_writes: vec![0; num_words].into_boxed_slice(),
+            page_erases: vec![0; num_pages].into_boxed_slice(),
+            interruption: Interruption::Ready,
+        };
+        assert!(buffer.is_word_aligned(buffer.options.page_size));
+        assert!(buffer.is_page_aligned(buffer.storage.len()));
+        buffer
+    }
+
+    /// Arms an interruption after a given delay.
+    ///
+    /// Before each subsequent mutable operation (write or erase), the delay is decremented if
+    /// positive. If the delay is elapsed, the operation is saved and an error is returned.
+    /// Subsequent operations will panic until the interrupted operation is [corrupted].
```

Comment:
Yes that's correct. This is to make sure an interrupted operation is handled and not forgotten. This helps get stronger confidence that the fuzzing code is correct. It also permits to be sure that the store reports all storage failures.

Reply:
The function that corrupts calls `get`, but technically you could also `disarm` or `reset`?

Reply:
Here is the state machine from user perspective:
- `write_slice`/`erase_page`: disarmed -> disarmed
- `arm_interruption`: disarmed -> armed
- `write_slice`/`erase_page`: armed -> armed (if delay remaining)
- `write_slice`/`erase_page`: armed -> triggered (if delay elapsed)
- `disarm_interruption`: armed -> disarmed
- `reset_interruption`: armed -> disarmed
- `corrupt_operation`: triggered -> disarmed
- `reset_interruption`: triggered -> disarmed

So I guess the answer to your question is like, you can use `reset` instead of `corrupt` or `disarm`, but that's it.

Reply:
Yeah, just wanted to point out that `corrupted` is not the only thing that stops subsequent operations from panicing.

Reply:
Oh good point. Yes I should add it.